feat(catalog): add support for entity policies in the new backend.

sennyeya opened the pull request:
## Hey, I just made a Pull Request!

closes #15455, pretty straightforward, I think it's worth keeping it with the rest of the `builder.add...` properties for now.

#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

Rugvip reviewed (commented):
Wanna highlight the proposal in https://github.com/backstage/backstage/issues/15455#issuecomment-1453279438, to keep this in a separate `CatalogModelExtensionPoint` (name TBD). It's perhaps slightly different and something to evolve separately from the way processors and providers are installed? In particular also because I think more methods might be needed than just adding new policies.

sennyeya commented:
@Rugvip Updated to a separate `CatalogModelExtensionPoint` with `addPolicies` and `replacePolicies`. Are there any other chunks we may want to include in this work?

freben commented:
I have a bit of a discussion point. What should we aim for as the long term goal for this?

Context: One of the key emerging things in the new backend system is that we are trying to start backends _as empty as is possible_. Ie having as few defaults as possible, and making it an intentional thing to explicitly add things to it. Hence one can add a new `addDefaults?: boolean` parameter to an old `createRouter` for example, so that the new backend plugin code can ensure it gets created stripped clean.

I think the policies generally fall in these categories now:

1. Actually important ones
   - `SchemaValidEntityPolicy` ensures that the entity is even remotely the right shape. But interestingly, this is already done by the catalog backend [far, far earlier](https://github.com/backstage/backstage/blob/72341bc0832a12f69a50fa6f90647bd6661be6a2/plugins/catalog-backend/src/processing/DefaultCatalogProcessingOrchestrator.ts#L132) so the check is sort of moot. Well, unless a processor completely borks the entity shape in its pre step, but that gets caught anyway by the validation step later.
   - `DefaultNamespaceEntityPolicy` makes sure that `metadata.namespace` always gets the value `default` if it's unset. This should honestly just be a hard coded thing in the catalog backend anyway.
   - `FieldFormatEntityPolicy` is the recipient for the field format validators, which is an odd indirection. To me it sounds like the new `CatalogModelExtensionPoint` (which I think is a great direction!) should have methods for directly setting field format rules instead of via replacing the policy.
1. Some very dubious ones
   - `GroupDefaultParentEntityPolicy` isn't even used
   - `NoForeignRootFieldsEntityPolicy` well yeah sure ... but again that should be a dedicated method on the `CatalogModelExtensionPoint` then
1. User defined ones. I think we should outright not allow this use case. Possibly we could just allow `add` and immediately marking it as deprecated and telling people to please not do that, or we could offer a policy-to-processor converter function that they can call as a compat feature while they migrate, so they can add them as processors instead.

As mentioned, I still think this particular extension point with this exact name is an awesome thing. If it has some specific great methods for doing model adjustments that are better than "just figure out how to write an entire processor", I think this can be a winner. But - what to do about policies in particular?

I vote for making a policy-to-processor converter function. Then we move toward stripping down the set of policies that are installed by default to an absolute minimum in the new backend system, and figure out how to move off of them entirely. What do you think?

sennyeya commented:
@freben Appreciate the in depth comment! Definitely agree, I've updated this PR to make `addPolicies` deprecated on release, added a new `policyToProcessorTransformer`, and added support for field validators on the extension point. I think the combination of `setFieldValidators` and `addPolicies` will cover the existing use cases for policies (minus removing one of the default policies, but I think that fits into "don't do this" territory).

freben commented on plugins/catalog-backend/api-report.md:
```diff
@@ -430,6 +430,11 @@ export type PlaceholderResolverResolveUrl = (
 ) => string;
 
+// @public
+export function policyToProcessorTransformer(
```
small nit i guess - a more active wording like `transformLegacyPolicyToProcessor` or something might read better

sennyeya commented:
@freben True, calling it out more clearly in the release notes/as a guide instead of introducing it as deprecated feels more intuitive. There's no current support and all we're saying it that there will be no support going forward as well.

freben commented:
@sennyeya needs a rebase now it seems

freben commented:
Oh and it also needs a changeset for -node still

freben commented:
interestingly, now i have to approve workflow runs what with the new username and all

aramissennyeydd commented:
@freben Had the same issue with #23036, I'm hoping this is just because it's the first few commits to the project on this account.

freben reviewed (approved):
Nice! Got a verbal ok from catalog maintainers for this too, let's 🚢
